File: src/lib/pjecalc/core/base/comum/logico-fuzzy.ts
/**
 * Porte 1:1 de LogicoFuzzy.java (41 linhas).
 *
 * Generalização de Boolean com exceções por período. Usado para perguntas como
 * "sábado é dia útil?" onde a resposta default pode ter exceções por período
 * (ex: em certa faixa de datas, sábado vira dia não-útil).
 *
 * Ref: pjecalc-fonte/base/br/jus/trt8/pjecalc/base/comum/LogicoFuzzy.java
 */
import type { Periodo } from './periodo';

export interface PeriodoParaExcecao {
  getPeriodo(): Periodo | null;
}

export class LogicoFuzzy<E extends PeriodoParaExcecao = PeriodoParaExcecao> {
  private considerar: boolean;
  private excecoes: Set<E> | null;

  static readonly VERDADEIRO: LogicoFuzzy = new LogicoFuzzy(true);
  static readonly FALSO: LogicoFuzzy = new LogicoFuzzy(false);

  constructor(considerar: boolean, excecoes: Set<E> | null = null) {
    this.considerar = considerar;
    this.excecoes = excecoes;
  }

  /**
   * Verifica se o valor lógico é verdadeiro para esta data,
   * invertendo-o se cair dentro de alguma exceção.
   *
   * Ref: LogicoFuzzy.java:30-40
   */
  isValido(date: Date): boolean {
    let resultado = this.considerar ?? false;
    if (this.excecoes !== null) {
      for (const excecao of this.excecoes) {
        const periodo = excecao.getPeriodo();
        if (periodo !== null && periodo.isPeriodoContemEsta(date)) {
          resultado = !resultado;
          break;
        }
      }
    }
    return resultado;
  }

  getConsiderar(): boolean { return this.considerar; }
  getExcecoes(): Set<E> | null { return this.excecoes; }
}

File: pjecalc-fonte/base/br/jus/trt8/pjecalc/base/comum/LogicoFuzzy.java
/*
 * Decompiled with CFR 0.152.
 */
package br.jus.trt8.pjecalc.base.comum;

import br.jus.trt8.pjecalc.base.comum.HelperDate;
import br.jus.trt8.pjecalc.base.comum.Utils;
import br.jus.trt8.pjecalc.base.comum.api.PeriodoParaExcecao;
import java.io.Serializable;
import java.util.Set;

public class LogicoFuzzy<E extends PeriodoParaExcecao>
implements Serializable {
    private static final long serialVersionUID = 118855806017405280L;
    private Boolean considerar;
    private Set<E> excecoes;
    public static final LogicoFuzzy<?> VERDADEIRO = new LogicoFuzzy(true);
    public static final LogicoFuzzy<?> FALSO = new LogicoFuzzy(false);

    public LogicoFuzzy(Boolean considerar, Set<E> excecoes) {
        this.considerar = considerar;
        this.excecoes = excecoes;
    }

    public LogicoFuzzy(Boolean considerar) {
        this.considerar = considerar;
        this.excecoes = null;
    }

    public boolean isValido(HelperDate date) {
        Boolean resultado = Utils.nulo(this.considerar) ? false : this.considerar;
        if (Utils.naoNulo(this.excecoes)) {
            for (PeriodoParaExcecao excecao : this.excecoes) {
                if (!Utils.naoNulo(excecao.getPeriodo()) || !excecao.getPeriodo().isPeriodoContemEsta(date.getDate())) continue;
                resultado = resultado == false;
                break;
            }
        }
        return resultado;
    }
}

